irods_specific_implementation/irods_rest_logical_path_api_implementation.h

```diff
@@ -0,0 +1,132 @@
+            }
+
+        }
+
+        std::tuple<Pistache::Http::Code, std::string>
+        delete_dispatcher(const Pistache::Rest::Request& _request,
+                   Pistache::Http::ResponseWriter& _response)
+        {
+            namespace fs = irods::experimental::filesystem;
+            namespace fscli = irods::experimental::filesystem::client;
+
+            try {
+                const auto _logical_path = _request.query().get("logical-path").get();
+                const auto _no_trash     = _request.query().get("no-trash").getOrElse("0");
+                const auto _recursive    = _request.query().get("recursive").getOrElse("0");
+                const auto _unregister   = _request.query().get("unregister").getOrElse("0");
+
+                const bool no_trash      = to_bool(_no_trash);
+                const bool recursive     = to_bool(_recursive);
+                const bool unregister    = to_bool(_unregister);
+
+                auto conn = get_connection(_request.headers().getRaw("authorization").value());
+
+                const fs::path logical_path = decode_url( _logical_path );
+
+                if ( ! recursive && fscli::is_collection(fscli::status(*conn(), logical_path)) )
+                {
+                    return make_error_response(
+                        USER_INCOMPATIBLE_PARAMS,
+                        "'recursive=1' required to delete a collection. Make sure you want to delete the whole sub-tree."
+                    );
+                }
+
+                fs::extended_remove_options opts {
+                    .no_trash      = no_trash,
+                    .verbose       = false,
+                    .progress      = false,
+                    .recursive     = recursive,
+                    .unregister    = unregister
+                };
+
+                bool status  = fs::client::remove(*conn(), logical_path, opts);
+
+                if ( ! status ) {
+                    return make_error_response(
+                        SYS_UNKNOWN_ERROR,
+                        "Client experienced an unknown error while processing your delete request."
+                    );
+                }
+
+                return std::make_tuple(
+                   Pistache::Http::Code::Ok,
+                   ""
+                );
+            }
+            catch (const fs::filesystem_error& e) {
+                error("Caught exception - [error_code={}] {}", e.code().value(), e.what());
+                return make_error_response(e.code().value(), e.what());
+            }
+            catch (const irods::exception& e) {
+                error("Caught exception - [error_code={}] {}", e.code(), e.what());
+                return make_error_response(e.code(), e.what());
+            }
+            catch (const std::exception& e) {
+                error("Caught exception - {}", e.what());
+                return make_error_response(SYS_INTERNAL_ERR, e.what());
+            }
```

Comment:
Perhaps worth considering a `catch(...)` for these as well just in case, e.g.
```c++
catch(...) {
    error("Caught unknown exception");
    return make_error_response(SYS_UKNOWN_ERROR, "Unknown error occurred");
}

Reply:
When is it a good idea to apply a catch-all? In fact, I guess my question is when it would _not_ be appropriate to add some kind of catch-all logic.

Reply:
It depends on the requirements.

Answering why you may want to use it is a lot easier than why you wouldn't. So here's a few things to consider:
- A third-party dependency could throw things that aren't derived from `std::exception`
- The implementation is required to not leak any exceptions
- You're unsure whether all cases are covered
    - It protects the user and also alerts us that we need to potentially add a new catch block
- You don't care what the error is, but you need to handle the exceptional case
    - This is a situation in which the code is expected to throw frequently

It is just another tool that is available to you.

Reply:
I think I asked elsewhere, but before I push a change that addresses this, what are general feels on std::current_exception within iRODS?

Reply:
We've never used it or needed it.

What's the motivation behind using it?

Reply:
Just intuitively, it seems like even if we're in the catch-all case we might want to capture the error message emitted by the exception.

Reply:
I think it may be worth an investigation, so perhaps we can capture that in an issue (maybe even in irods/irods). However, the main motivation for my suggestion is bullet 2 from @korydraughn's comment:

> The implementation is required to not leak any exceptions

If we can show that something further up the call stack is handling the catch-all case, then it is not necessary here, IMO.

Reply:
Correct.

> Just intuitively, it seems like even if we're in the catch-all case we might want to capture the error message emitted by the exception.

This assumes we know what the exception's type is. If we're catching all exception types we know about, what do you expect `std::current_exception` to point to?

See the example here: https://en.cppreference.com/w/cpp/error/current_exception